File: Array/easy/second_largest_element_in_an_array_without_sorting.cpp
//
// Created by santosh on 12/5/25.
//

#include<bits/stdc++.h>
using namespace std;

int secondLargestElement(vector<int>& nums)
{
    int smallest = INT_MAX;
    int second_smallest = -1;
    int largest = INT_MIN;
    int second_largest = -1;
    for (int i = 0; i < nums.size(); i++)
    {
        // largest and second largest
        if (nums[i] > largest)
        {
            second_largest = largest;
            largest = nums[i];
        }
        else if (nums[i] < largest && nums[i] > second_largest)
        {
            second_largest = nums[i];
        }
        // smallest and second smallest

        if (nums[i] < smallest)
        {
            second_smallest = smallest;
            smallest = nums[i];
        }
        if (nums[i] > smallest && nums[i] < second_smallest)
        {
            second_smallest = nums[i];
        }
    }

}

int main()
{
    vector<int> nums = {
        3, 4, 7, 7, 5, 3, 0,2
    };
    int second_largest = secondLargestElement(nums);
    cout << second_largest << endl;
}

File: Array/easy/second_largest_element_in_an_array_without_sorting.py
# Second largest and second smallest element without sorting
# Created by santosh

def second_largest_element(nums: list[int]) -> dict:
    smallest = float('inf')
    second_smallest = float('inf')
    largest = float('-inf')
    second_largest = float('-inf')

    for num in nums:
        # Largest and second largest
        if num > largest:
            second_largest = largest
            largest = num
        elif num < largest and num > second_largest:
            second_largest = num

        # Smallest and second smallest
        if num < smallest:
            second_smallest = smallest
            smallest = num
        elif num > smallest and num < second_smallest:
            second_smallest = num

    return {
        "largest": largest,
        "second_largest": second_largest,
        "smallest": smallest,
        "second_smallest": second_smallest,
    }


if __name__ == "__main__":
    nums = [3, 4, 7, 7, 5, 3, 0, 2]
    result = second_largest_element(nums)
    print("Second largest:", result["second_largest"])
    print("Second smallest:", result["second_smallest"])
